File: 2022/2022W10/preppindataW10.R
library(readxl)
library(dplyr)
library(xml2)
library(stringr)

xlsx <- "C:/Data/PreppinData/PD Bechdel Test.xlsx"

final <- read_excel(xlsx, 'Webscraping') %>% 
  rowwise() %>%
  mutate('Movie' = str_extract(`DownloadData`, '(?<=>)(([^><])+)(?=<\\/a>)'),
         'Movie' = xml_text(read_html(charToRaw(`Movie`))),
         'Categorisation' = str_extract(`DownloadData`, '(?<=title="\\[)(.*)(?=\\])'),
         'Ranking' = case_when(
           `Categorisation`=='There are two or more women in this movie and they talk to each other about something other than a man' ~ 1,
           `Categorisation`=='There are two or more women in this movie and they talk to each other about something other than a man, although dubious' ~ 2,
           `Categorisation`=='There are two or more women in this movie, but they only talk to each other about a man' ~ 3,
           `Categorisation`=="There are two or more women in this movie, but they don't talk to each other" ~ 4,
           TRUE ~ 5),
         'Pass/Fail' = if_else(`Ranking`<=2, 'Pass', 'Fail')) %>%
  group_by(`Movie`, `Year`) %>%
  filter(`Ranking` == max(`Ranking`)) %>%
  select(c('Movie', 'Year', 'Pass/Fail', 'Ranking', 'Categorisation')) %>%
  distinct()

View(final)

File: 2022/2022W10/PreppindataW10.py
import pandas as pd
import html
import re

xlsx = pd.ExcelFile(r"C:\Data\PreppinData\PD Bechdel Test.xlsx")

ranking = {"There are two or more women in this movie and they talk to each other about something other than a man":1,
           "There are two or more women in this movie and they talk to each other about something other than a man, although dubious":2,
           "There are two or more women in this movie, but they only talk to each other about a man":3,
           "There are two or more women in this movie, but they don't talk to each other":4,
           "Fewer than two women in this movie":5 }

movie = pd.read_excel(xlsx,'Webscraping')
movie['Movie'] = movie['DownloadData'].apply(lambda x: html.unescape(re.search('(?<=>)(([^><])+)(?=<\/a>)', x).group(1)))
movie['Categorisation'] = movie['DownloadData'].apply(lambda x: re.search('(?<=title="\[)(.*)(?=\])', x).group(1))
movie['Ranking'] = movie['Categorisation'].apply(lambda x: ranking.get(x))
movie['Worst Ranking'] = movie['Ranking'].groupby([movie['Movie'], movie['Year']]).transform('max')
movie = movie[movie['Ranking'] == movie['Worst Ranking']].drop('DownloadData', axis=1).drop_duplicates()
movie['Pass/Fail'] = movie['Ranking'].apply(lambda x: 'Pass' if x<=2 else 'Fail')
movie = movie[['Movie', 'Year', 'Pass/Fail', 'Ranking', 'Categorisation']]
